git_machete/cmd.py

```diff
@@ -1978,39 +1978,55 @@ def run_post_slide_out_hook(cli_ctxt: CommandLineContext, new_upstream: str, sli
         if not new_upstream:
             raise MacheteException(f"No upstream branch defined for `{b}`, cannot slide out")
-        dbs = down_branches.get(b)
+
+    # Verify that all "interior" slide-out branches have a single downstream pointing to the next slide-out
+    for bu, bd in zip(branches_to_slide_out[:-1], branches_to_slide_out[1:]):
+        dbs = down_branches.get(bu)
         if not dbs or len(dbs) == 0:
-            raise MacheteException(f"No downstream branch defined for `{b}`, cannot slide out")
+            raise MacheteException(f"No downstream branch defined for `{bu}`, cannot slide out")
         elif len(dbs) > 1:
             flat_dbs = ", ".join(f"`{x}`" for x in dbs)
-            raise MacheteException(f"Multiple downstream branches defined for `{b}`: {flat_dbs}; cannot slide out")
+            raise MacheteException(f"Multiple downstream branches defined for `{bu}`: {flat_dbs}; cannot slide out")
+        elif dbs != [bd]:
+            raise MacheteException(f"'{bd}' is not downstream of '{bu}', cannot slide out")
 
-    for bu, bd in zip(branches_to_slide_out[:-1], branches_to_slide_out[1:]):
         if up_branch[bd] != bu:
             raise MacheteException(f"`{bu}` is not upstream of `{bd}`, cannot slide out")
 
+    # Get new branches
     new_upstream = up_branch[branches_to_slide_out[0]]
-    new_downstream = down_branches[branches_to_slide_out[-1]][0]
+    new_downstreams = down_branches.get(branches_to_slide_out[-1], [])
+
+    # Remove the slide-out branches from the tree
     for b in branches_to_slide_out:
         up_branch[b] = None
         down_branches[b] = None
+    down_branches[new_upstream] = [b for b in down_branches[new_upstream] if b != branches_to_slide_out[0]]
 
-    go(cli_ctxt, new_downstream)
-    up_branch[new_downstream] = new_upstream
-    down_branches[new_upstream] = [(new_downstream if x == branches_to_slide_out[0] else x) for x in down_branches[new_upstream]]
+    # Reconnect the downstreams to the new upstream in the tree
+    for new_downstream in new_downstreams:
+        up_branch[new_downstream] = new_upstream
+        down_branches[new_upstream].append(new_downstream)
+
+    # Update definition, fire post-hook, and perform the branch update
     save_definition_file()
-    run_post_slide_out_hook(cli_ctxt, new_upstream, branches_to_slide_out[-1], [new_downstream])
-    if cli_ctxt.opt_merge:
-        print(f"Merging {bold(new_upstream)} into {bold(new_downstream)}...")
-        merge(cli_ctxt, new_upstream, new_downstream)
-    else:
-        print(f"Rebasing {bold(new_downstream)} onto {bold(new_upstream)}...")
-        rebase(cli_ctxt, f"refs/heads/{new_upstream}", cli_ctxt.opt_down_fork_point or fork_point(cli_ctxt, new_downstream, use_overrides=True), new_downstream)
+    run_post_slide_out_hook(cli_ctxt, new_upstream, branches_to_slide_out[-1], new_downstreams)
+
+    go(cli_ctxt, new_upstream)
+    for new_downstream in new_downstreams:
```

Comment:
Hmmm actually... for the UX point of view, I have a concern here: in case the rebase is interactive (so `-n` option is NOT passed), will the user be able to find out which downstream branch is currently getting rebased? 🤔 I'd just want to make sure that we don't leave the user confused with multiple rebases happening one by one

Reply:
Possibly making it similar to what happens in `git machete traverse` (so explicit questions like `Rebase branch 'foo' onto 'bar'?`) would be a good choice for slide-outs where >1 downstream is involved?

Reply:
🤔 The currently experience isn't bad. The `slide_out` function prints an update message before opening the interactive rebase, so there's pretty good messaging before the rebase fires. Eg:

```bash
mdemo$ git machete status
  main
  │
  └─slide_out
    │
    └─child_a
    │
    └─child_b
    
mdemo$ git machete slide-out
Rebasing child_a onto main...
Successfully rebased and updated refs/heads/child_a.
Rebasing child_b onto main...
Successfully rebased and updated refs/heads/child_b.

mdemo$ git machete status
  main
  │
  └─child_a
  │
  └─child_b
```

I don't think you want to prompt during the `slide-out` operation, because at that point the updated machete configuration has already been written. Aborting there would leave the tree in a vaguely inconsistent state that could be hard to back out from. 

I *personally* don't want an interactive prompt during the operation as slide-out as a primitive operation that operates on a single branch, rather than traverse which is *potentially* operating on multiple branches.

If I'm calling from the cli via `git machete slide-out` I've already looked at and selected the branch, and if it's run during a traverse then I've already been prompted to confirm the slide-out operation at a branch level before firing.

Reply:
Okay, you've convinced me :) let's resolve the confusion around `down-fork-point` tho